tests/helpers/deviceappproxy/deviceappbtc.py

```diff
@@ -0,0 +1,340 @@
+                    # Segwit consensus rules state that if an input from the tx to sign refers to a Segwit prev_tx,
+                    # then the input script to hash with that input shall be:
+                    # 19 || 76a914 || 20-byte pubkey hash from prev_tx's requested output || 88ac
+                    scripts.append(bytes.fromhex("76a914") + pubkey_hash_from_script(script_pubkey)
+                                   + bytes.fromhex("88ac"))
+                else:
+                    scripts.append(script_pubkey)
+            else:
+                raise NotImplementedError(f"Unsupported hashing mode provided: {mode}")
+
+        # version || input count
+        # Note: input_count is set to 1 when sending inputs individually with their script
+        version_chunk = parsed_tx.version.buf + cast(ZcashExtHeader, parsed_tx.header.ext).version_group_id.buf \
+            if mode.is_zcash_input_hash or mode.is_sapling_input_hash \
+            else parsed_tx.version.buf
+        payload_chunks = [
+            version_chunk + bytes.fromhex("01")
+            if mode.is_hash_with_script and mode.is_segwit_zcash_or_sapling_input_hash
+            else version_chunk + parsed_tx.input_count.buf
+        ]
+        # Compose a list of: input || script_len (possibly 0) || script (possibly None) || sequence_nb
+        for f_input, script in zip(formatted_inputs, scripts):
+            input_idx = self._get_input_index(parsed_tx, f_input, endianness)
+            # add input with or without input script, depending on hashing phase
+            if mode.is_segwit_zcash_or_sapling_input_hash:
+                if mode.is_hash_with_script:
+                    payload_chunks.extend(
+                        [  # [input||script_len, script||sequence]
+                            f_input + TxVarInt.to_bytes(len(script), 'little'),
+                            script + parsed_tx.inputs[input_idx].sequence_nb.buf
+                        ])
+                else:   # Hash inputs w/o scripts
+                    payload_chunks.extend(
+                        [  # [input||0 (no script), sequence]
+                            f_input + b'\x00', parsed_tx.inputs[input_idx].sequence_nb.buf
+                        ])
+            else:   # BTC or BCash tx
+                if script is not None:
+                    payload_chunks.extend(
+                        [  # [input||script_len, script||sequence]
+                            f_input + TxVarInt.to_bytes(len(script), 'little'),
+                            script + parsed_tx.inputs[input_idx].sequence_nb.buf
+                        ])
+                else:
+                    payload_chunks.extend(
+                        [  # [input||script_len (00), sequence]
+                            f_input + b'\x00', parsed_tx.inputs[input_idx].sequence_nb.buf
+                        ])
```

Comment:
the two branches of this if...else look pretty much identical

```suggestion
            with_script = (script is not None) or (mode.is_segwit_zcash_or_sapling_input_hash and mode.is_hash_with_script)
            if with_script:
                payload_chunks.extend(
                    [  # [input||script_len, script||sequence]
                        f_input + TxVarInt.to_bytes(len(script), 'little'),
                        script + parsed_tx.inputs[input_idx].sequence_nb.buf
                    ])
            else:
                payload_chunks.extend(
                    [  # [input||0 (no script), sequence]
                        f_input + b'\x00', parsed_tx.inputs[input_idx].sequence_nb.buf
                    ])
```

Reply:
Indeed. Why did I write it this way? The difference between the two branches is in what the script is made of, but that's already been taken care of when the code gets there. 
==> Code changed as suggested.

Reply:
indeed 👍